core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,404 @@
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import java.io.ByteArrayInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.exceptions.NotFoundException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.hadoop.HadoopConfigurable;
+import org.apache.iceberg.hadoop.HadoopFileIO;
+import org.apache.iceberg.hadoop.SerializableConfiguration;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.SerializableSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of FileIO that adds metadata content caching features.
+ * <p>
+ * This FileIO intend to speedup scan planning by caching most of the table metadata contents in memory and reduce
+ * remote reads.
+ */
+public class CachingFileIO implements FileIO, HadoopConfigurable {
+  private static final Logger LOG = LoggerFactory.getLogger(CachingFileIO.class);
+  private static ContentCache sharedCache;
+
+  private static ContentCache createCache(Map<String, String> properties) {
+    long durationMs = PropertyUtil.propertyAsLong(properties,
+        CatalogProperties.CONTENT_CACHE_EXPIRATION_INTERVAL_MS,
+        CatalogProperties.CONTENT_CACHE_EXPIRATION_INTERVAL_MS_DEFAULT);
+    long totalBytes = PropertyUtil.propertyAsLong(properties,
+        CatalogProperties.CONTENT_CACHE_MAX_TOTAL_BYTES,
+        CatalogProperties.CONTENT_CACHE_MAX_TOTAL_BYTES_DEFAULT);
+    long contentLength = PropertyUtil.propertyAsLong(properties,
+        CatalogProperties.CONTENT_CACHE_MAX_CONTENT_LENGTH,
+        CatalogProperties.CONTENT_CACHE_MAX_CONTENT_LENGTH_DEFAULT);
+
+    if (durationMs >= 0 && totalBytes > 0) {
+      return new ContentCache(durationMs, totalBytes, contentLength);
+    } else {
+      return null;
+    }
+  }
+
+  public static synchronized ContentCache getSharedCache(Map<String, String> properties) {
+    if (sharedCache == null) {
+      sharedCache = createCache(properties);
+    }
+    return sharedCache;
+  }
+
+  private SerializableSupplier<Configuration> hadoopConf;
+  private ContentCache fileContentCache;
+  private Map<String, String> properties;
+  private FileIO wrappedIO;
+
+  /**
+   * Constructor used for dynamic FileIO loading.
+   * <p>
+   * {@link Configuration Hadoop configuration} must be set through {@link HadoopFileIO#setConf(Configuration)}
+   */
+  public CachingFileIO() {
+  }
+
+  public CachingFileIO(Configuration hadoopConf) {
+    this(new SerializableConfiguration(hadoopConf)::get);
+  }
+
+  public CachingFileIO(SerializableSupplier<Configuration> hadoopConf) {
+    this.hadoopConf = hadoopConf;
```

Comment:
Why does this reference a Hadoop Configuration? It seems like this doesn't need to rely on Hadoop.

Reply:
Nevermind, I see this hasn't been refactored to pass in the inner `FileIO` yet.

Reply:
This is now removed.